METRON-1031: Management UI Cannot Start Topologies in Kerberized Environment

nickwallen opened the pull request:
The Metron Management UI cannot start topologies in a kerberized environment.  This includes the Parser, Enrichment, and Indexing topologies.

It seems that Metron REST does not pass "-ksp" to the start topology commands. As a result, topologies that are started with the Management UI cannot start a producer against a kerberized Kafka.

### Changes

* Altered the `StormCLIWrapper` to add logging of the actual CLI commands that are executed.
* The `-ksp` argument is always used (with Kerberos or not) when launching the Parser, Enrichment, and Indexing topologies.
* Added a Spring property `kafka.security.protocol` that serves as the `-ksp` argument when launching the topologies.
* Ambari sets this Spring property based on Kafka's own configuration in Ambari.

### Testing

To manually test this, validate that all of the topologies can be stopped and started using the Management UI and Swagger UI both before and after kerberization.

1. Launch the Full Dev environment.
1. Stop then start the Parser topologies using the Management UI.
1. Ensure that data is being consumed by the topologies and landing in the Elasticsearch indices.
1. Kerberize the Full Dev environment [using these instructions](https://github.com/apache/metron/blob/master/metron-deployment/Kerberos-ambari-setup.md).
1. Kerberize the Sensor Stubs.

    Add the following to somewhere near the top of `/etc/init.d/sensor-stubs`.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    ```
    Add the `--security-protocol`.
    ```
    sed -i 's/node1:6667 --topic/node1:6667 --security-protocol PLAINTEXTSASL --topic/' /opt/sensor-stubs/bin/start-*-stub
    ```
    Restart the stubs.
    ```
    for sensorstub in bro snort; do
        service sensor-stubs stop ${sensorstub};
        service sensor-stubs start ${sensorstub};
    done
    ```
1. Ensure data is being sent to the Parsers from the Sensor Stubs.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic bro
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic snort
    ```
1. Ensure that the Parsers are sending data to the Enrichment topology.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic enrichments
    ```
1. Stop the Bro & Snort topology using the Management UI.
1. Start the Bro topology using the Management UI.
1. Ensure that the Bro Parser is sending data to the Enrichment topology.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic enrichments
    ```


### Pull Request Checklist

- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

merrimanr commented on metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/StormCLIWrapper.java:
```diff
@@ -75,37 +81,50 @@ public int stopIndexingTopology(boolean stopNow) throws RestException {
 
   protected String[] getEnrichmentStartCommand() {
-    String[] command = new String[1];
+    String[] command = new String[3];
```
I don't think the -ksp flag is needed for the enrichment or indexing topologies.  This setting comes from enrichment.properties and elasticsearch.properties.

nickwallen replied:
Makes sense now.  I will remove -ksp from both Enrichment and Indexing start commands.  Thanks!

simonellistonball replied:
Has this changed since 0.4.0 RC1? Didn't work when I tried it until ksp added in the start script.

merrimanr replied:
No it's always been like that.  The parser topology start script is the only one that supports the ksp flag.  The enrichment and elasticsearch topology start scripts don't expect any input parameters.

simonellistonball replied:
right, my bad, it was just parser script I had to change

nickwallen commented:
- I removed `-ksp` from Enrichment and Indexing.
- I added `-e` also to the Parsers.  This is also required to get them running.
- Updated the test plan.

I ran through a full test once successfully, but I changed a few minor things and just need to run through the manual test once more.

nickwallen commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/templates/metron.j2:
```diff
@@ -37,3 +37,5 @@ SECURITY_ENABLED={{security_enabled|lower}}
 {% if metron_keytab_path is defined %}METRON_SERVICE_KEYTAB="{{metron_keytab_path}}"
 {% endif %}
+KAFKA_SECURITY_PROTOCOL="{{kafka_security_protocol}}"
+PARSER_TOPOLOGY_OPTIONS="/home/metron/.storm/storm.config"
```
I did not find any existing property that defines the `PARSER_TOPOLOGY_OPTIONS` value.  It is hard coded in one of the MPack Python scripts (`parser_commands.py`) as `' -e ~' + self.__params.metron_user + '/.storm/storm.config'`.

I thought that if I just put the exact value here, we are half way to making it configurable in Ambari, if we choose to do so.

I also tried to use `~${METRON_HOME}/.storm/storm.config`, but the `ProcessBuilder` does not do tilde expansion for user's home directories.

merrimanr replied:
I think that's fine.  If you really wanted it to mirror the implementation in the MPack, you could do:
`PARSER_TOPOLOGY_OPTIONS="/home/{{metron_user}}/.storm/storm.config"`

nickwallen commented:
@merrimanr  Ran it through a full test as described in the PR description and its looking good.  Please take a look when you can.

merrimanr commented:
+1 validated with instructions in the PR description.  Nice work!
